tools/file_packager.py

```diff
@@ -475,53 +485,60 @@ def was_seen(name):
+      'audio': '1' if filename[-4:] in AUDIO_SUFFIXES else '0',
+    })
   else:
     assert 0
 
 if has_preloaded:
-  # Get the big archive and split it up
-  if no_heap_copy:
-    use_data = '''
-      // copy the entire loaded file into a spot in the heap. Files will refer to slices in that. They cannot be freed though
-      // (we may be allocating before malloc is ready, during startup).
-      var ptr = Module['getMemory'](byteArray.length);
-      Module['HEAPU8'].set(byteArray, ptr);
-      DataRequest.prototype.byteArray = Module['HEAPU8'].subarray(ptr, ptr+byteArray.length);
-'''
+  if not lz4:
+    # Get the big archive and split it up
+    if no_heap_copy:
+      use_data = '''
+        // copy the entire loaded file into a spot in the heap. Files will refer to slices in that. They cannot be freed though
+        // (we may be allocating before malloc is ready, during startup).
+        var ptr = Module['getMemory'](byteArray.length);
+        Module['HEAPU8'].set(byteArray, ptr);
+        DataRequest.prototype.byteArray = Module['HEAPU8'].subarray(ptr, ptr+byteArray.length);
+  '''
+    else:
+      use_data = '''
+        // Reuse the bytearray from the XHR as the source for file reads.
+        DataRequest.prototype.byteArray = byteArray;
+  '''
+    use_data += '''
+          var files = metadata.files;
+          for (i = 0; i < files.length; ++i) {
+            DataRequest.prototype.requests[files[i].filename].onload();
+          }
+    '''
+    use_data += "          Module['removeRunDependency']('datafile_%s');\n" % data_target
+
+    if Compression.on:
+      use_data = '''
+        Module["decompress"](byteArray, function(decompressed) {
+          byteArray = new Uint8Array(decompressed);
+          %s
+        });
+      ''' % use_data
   else:
+    # LZ4FS usage
+    temp = data_target + '.orig'
+    shutil.move(data_target, temp)
+    meta = run_js(shared.path_from_root('tools', 'lz4-compress.js'), shared.NODE_JS, [shared.path_from_root('src', 'mini-lz4.js'), temp, data_target], stdout=PIPE)
+    os.unlink(temp)
     use_data = '''
-      // Reuse the bytearray from the XHR as the source for file reads.
-      DataRequest.prototype.byteArray = byteArray;
-'''
-  for file_ in data_files:
-    if file_['mode'] == 'preload':
-      use_data += '          DataRequest.prototype.requests["%s"].onload();\n' % (escape_for_js_string(file_['dstpath']))
-  use_data += "          Module['removeRunDependency']('datafile_%s');\n" % data_target
-
-  if Compression.on:
-    use_data = '''
-      Module["decompress"](byteArray, function(decompressed) {
-        byteArray = new Uint8Array(decompressed);
-        %s
-      });
-    ''' % use_data
+          var compressedData = %s;
+          compressedData.data = byteArray;
+          assert(typeof LZ4 === 'object', 'LZ4 not present - does your app call emscripten_init_lz4(), which should ensure LZ4 is linked in?');
```

Comment:
`loadPackage` already calls `LZ4.init()`, so C code having t all `emscripten_init_lz4` is just to escape DCE kicking in? If we used a linker flag `-llz4.js` or similar, then the C code would not need to be changed?